src/server/json_family.cc

```diff
@@ -964,73 +917,74 @@ OpResult<vector<OptString>> OpMGet(const OpArgs& op_args, const vector<string_vi
 OpResult<vector<OptSizeT>> OpFields(const OpArgs& op_args, string_view key,
                                     JsonExpression expression) {
-  OpResult<json> result = GetJson(op_args, key);
+  OpResult<JsonType*> result = GetJson(op_args, key);
   if (!result) {
     return result.status();
   }
 
   vector<OptSizeT> vec;
-  auto cb = [&vec](const string_view& path, const json& val) {
+  auto cb = [&vec](const string_view& path, const JsonType& val) {
     vec.emplace_back(CountJsonFields(val));
   };
-
-  expression.evaluate(*result, cb);
+  const JsonType& json_entry = *(result.value());
+  expression.evaluate(json_entry, cb);
   return vec;
 }
 
 // Returns json vector that represents the result of the json query.
-OpResult<vector<json>> OpResp(const OpArgs& op_args, string_view key, JsonExpression expression) {
-  OpResult<json> result = GetJson(op_args, key);
+OpResult<vector<JsonType>> OpResp(const OpArgs& op_args, string_view key,
+                                  JsonExpression expression) {
+  OpResult<JsonType*> result = GetJson(op_args, key);
   if (!result) {
     return result.status();
   }
 
-  vector<json> vec;
-  auto cb = [&vec](const string_view& path, const json& val) { vec.emplace_back(val); };
-
-  expression.evaluate(*result, cb);
+  vector<JsonType> vec;
+  auto cb = [&vec](const string_view& path, const JsonType& val) { vec.emplace_back(val); };
+  const JsonType& json_entry = *(result.value());
+  expression.evaluate(json_entry, cb);
   return vec;
 }
 
 // Returns boolean that represents the result of the operation.
 OpResult<bool> OpSet(const OpArgs& op_args, string_view key, string_view path,
-                     string_view json_str) {
-  // Check if the supplied JSON is valid.
-  optional<json> j = ConstructJsonFromString(json_str);
-  if (!j) {
+                     std::string_view json_str) {
+  std::optional<JsonType> parsed_json = JsonFromString(json_str);
+  if (!parsed_json) {
+    LOG(WARNING) << "got invalid JSON string '" << json_str << "' cannot be saved";
     return OpStatus::SYNTAX_ERR;
   }
 
   // The whole key should be replaced.
+  // NOTE: unlike in Redis, we are overriding the value when the path is "$"
+  // this is regardless of the current key type. In redis if the key exists
+  // and its not JSON, it would return an error.
   if (path == "." || path == "$") {
-    SetString(op_args, key, j->as_string());
+    SetJson(op_args, key, std::move(parsed_json.value()));
     return true;
   }
 
   // Update the existing JSON.
-  OpResult<json> result = GetJson(op_args, key);
+  OpResult<JsonType*> result = GetJson(op_args, key);
   if (!result) {
     return result.status();
   }
 
-  json existing_json = move(result.value());
   bool path_exists = false;
-  auto cb = [&](const string& path, json& val) {
+  auto cb = [&, new_json = std::move(parsed_json.value())](const string& path, JsonType& val) {
     path_exists = true;
-    val = *j;
+    val = std::move(new_json);
```

Comment:
Please check all the lambdas, you fell for a common trap 🙂 

Lambdas are immutable by default, this also applied for referencing its members.

```c++
#include <iostream>

using namespace std;

struct S {
    S() = default;
    S(const S& o) { cout << "COPY" << endl; }
    S(S&& o) { cout << "MOVE" << endl; }

    void operator=(const S& o) { cout << "COPY" << endl; }
    void operator=(S&& o) { cout << "MOVE" << endl;}
};

int main() {
    S out;

    S cur;
    auto cb1 = [ns = std::move(cur)](S& out) mutable {
        out = std::move(ns);
    };
    cb1(out);

    cout << "---" << endl;

    auto cb2 = [ns = std::move(cur)](S& out) {
        out = std::move(ns);
    };
    cb2(out);
}
```

See, this prints 
```
MOVE
MOVE
---
MOVE
COPY
```

with a copy in the non mutable case

Reply:
Thanks - this is the only place where this is done.

Reply:
There is a reason for the copy (though you're right about you're comment)
In the case that we're updating an existing value, we cannot move out of the new json object. This is because we may apply this new json value multiple times. For example if we are changing an existing array with 10 elements - for example [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], and we replacing them with 101 then the input for this is just "101" as the JSON that we're parsing and this JSON value is applied for each element in the array so we will end with [101, 101, 101, 101, 101, 101, 101, 101, 101, 101]
If we will move out, then after the first call, there will be no value for what was 101.
I've changed the code and added more comments to make it clear that this is what's happening.